### PR TITLE
WIP: Support Redshift groups in grant statements

### PR DESCRIPTION
**Pull request by y-trobinso**
Prototype support for Redshift groups within grant statements. Rationale follows.

Users work as normal, but it seems Redshift does not support groups in dialect of postgresql system information functions (eg, https://docs.aws.amazon.com/redshift/latest/dg/r_HAS_TABLE_PRIVILEGE.html). As such, we have to take a deep dive into [pg_class](https://www.postgresql.org/docs/8.0/static/catalog-pg-class.html)'s `aclitem[]` framework, checking for ACLs connected to `group <groupname>` in the underlying stringlike structure (eg, https://docs.aws.amazon.com/redshift/latest/dg/r_PG_DEFAULT_ACL.html).

This is, needless to say, elaborate and extremely messy. I'm open to better ways to extract this information if they exist, but I do not presently see these in Redshift's documentation.

### CODE REVIEW

**Comment by y-trobinso** [comment]
Unfortunately, the custom UNLESS syntax uses several unsupported commands in Redshift (https://docs.aws.amazon.com/redshift/latest/dg/c_unsupported-postgresql-functions.html). I'll see if I can make these work using Puppet regular expression tools instead. If not, we'll need to shelve this until Redshift rebases its servers onto postgresql >= 8.4.

File: manifests/server/grant.pp [hunk 1]
```diff
@@ -302,18 +327,63 @@
...
+  if ($dialect == 'redshift' and $_lowercase_role =~ /^group (.*)/) {
+    # Built-in functions such as has_table_privilege don't work on
+    # groups in Redshift at this writing. Similarly,
+    # information_schema role tables do not appear to be consistently
+    # kept up to date. As such, we have to dive into the low-level
+    # aclitem[] within pg_catalog.pg_class to find what we're looking
+    # for.
+    # 
+    # This only works with object types that cleanly map to
+    # information_schema, and is incompatible with complex permission
```
Comment (danielpops) [inline manifests/server/grant.pp]: can you give an example of a complex permission type that wouldn't work here, versus one that would?

File: manifests/server/grant.pp [hunk 2]
```diff
@@ -302,18 +327,63 @@
...
       # Never put double quotes into has_*_privilege function
       $_granted_object = join($_object_name, '.')
+      $_schema = $_object_name[0]
+      $_relation = $_object_name[1]
     }
     default: {
       $_granted_object = $_object_name
       $_togrant_object = $_object_name
+      $_schema = $_object_name
+      $_relation = '%'
```
Comment (danielpops) [inline manifests/server/grant.pp:324]: disclaimer: i don't know anything about postgres :)

What does relation = '%' mean? all relations?

Feel free to be very terse here. Mostly just curious, and my google-fu couldn't find the answer.

File: manifests/server/grant.pp [hunk 3]
```diff
@@ -337,8 +407,13 @@
     require          => Class['postgresql::server']
   }
 
-  if($role != undef and defined(Postgresql::Server::Role[$role])) {
-    Postgresql::Server::Role[$role]->Postgresql_psql["${title}: grant:${name}"]
+  if($role != undef) {
+    if ($_lowercase_role =~ /^group (.*)/ and defined(Postgresql::Server::Dbgroup["#$1"])) {
+      Postgresql::Server::Dbgroup["#$1"]->Postgresql_psql["${title}: grant:${name}"]
+    }
+    if defined(Postgresql::Server::Role[$role]) {
```
Comment (danielpops) [inline manifests/server/grant.pp]: why the `if defined`? shouldn't the role already exist? If it doesn't this should be an error, right?

File: manifests/server/grant.pp [hunk 4]
```diff
@@ -323,7 +393,7 @@
   }
 
   $grant_cmd = "GRANT ${_privilege} ON ${_object_type} \"${_togrant_object}\" TO
-      \"${role}\""
+      ${role}"
```
Comment (danielpops) [inline manifests/server/grant.pp:381]: is it safe to do this? can the role have spaces?

File: manifests/server/grant.pp [hunk 5]
```diff
@@ -302,18 +327,63 @@
...
+    from
+    (SELECT c.relacl FROM pg_class c
+     left join pg_namespace nsp on (c.relnamespace = nsp.oid)
+    WHERE
+     nsp.nspname = '${_schema}'
+     AND c.relname = '${_relation}'
+     AND c.reltype = '${_lowercase_object_type}'
+    ))) > 0"
+    if $_lowercase_object_type == "all tables in schema" {
+      warning('UNLESS clause is not yet supported for ALL TABLES IN SCHEMA granted to groups')
```
Comment (danielpops) [inline manifests/server/grant.pp]: can you include a comment with a github issue or stackoverflow etc of why this isn't supported and/or what would be pending for it to be supported?

File: manifests/server/role.pp [hunk 6]
```diff
@@ -206,6 +206,7 @@
           environment => $environment,
         }
       } elsif ($dialect == 'redshift') {
+        warning('UNLESS clause support is not yet supported for setting password on Redshift users')
```
Comment (danielpops) [inline manifests/server/role.pp]: Same comment. Can you include a comment with a pointer to what specifically blocks this support?

File: manifests/server/grant.pp (hunk 2 above)
Comment (y-trobinso) [reply manifests/server/grant.pp:324]: % is a wildcard character in ANSI SQL described here: https://www.w3schools.com/sql/sql_wildcards.asp. Postgres, like its brethren, is a dialect of ANSI SQL.

I'm still mid-testing to verify if equivalence will work or a LIKE query will be required. If the latter is required, that will change here.

File: manifests/server/grant.pp (hunk 1 above)
Comment (y-trobinso) [reply manifests/server/grant.pp]: I'll reword this. Basically, ALL TABLES IN SCHEMA and ALL SEQUENCES IN SCHEMA will not work here (doubly so for the latter, as Redshift does not support sequences).

File: manifests/server/grant.pp (hunk 5 above)
Comment (y-trobinso) [reply manifests/server/grant.pp]: https://docs.aws.amazon.com/redshift/latest/dg/c_unsupported-postgresql-functions.html describes this in detail. I'll link it in a comment.

File: manifests/server/grant.pp (hunk 4 above)
Comment (y-trobinso) [reply manifests/server/grant.pp:381]: This is explicit to support `group <groupname>` in the role parameter. It seems Redshift explicitly supports spaces in the role name for this purpose, as can be seen in the aclitem[] example here: https://docs.aws.amazon.com/redshift/latest/dg/r_PG_DEFAULT_ACL.html

So, yes, spaces are supported. And while I'm nervous about removing string quoting here, it's required to support mixed use of including the group keyword, unless we opt to provide that as a separate variable. Quotation of the username or group is behavior that is, furthermore, not specified in https://docs.aws.amazon.com/redshift/latest/dg/r_GRANT.html

tl;dr, this is explicit, and since we're evaluating local code, should be safe. I can add a warning or tokenize this for safety if we can define scenarios in which misuse or injection (by user input, as opposed to puppet logic) will exist.

File: manifests/server/grant.pp (hunk 3 above)
Comment (y-trobinso) [reply manifests/server/grant.pp]: This is a holdover from the original module. I would assume this is to allow the role to be defined independently of the grant (ie, in later reviews), and to specify an ordering when (and only when) it exists (since puppet will topologically sorts class evaluation instead of compiling each class in evaluation order).

For that reason, this seems fine to me.

File: manifests/server/role.pp (hunk 6 above)
Comment (y-trobinso) [reply manifests/server/role.pp]: I describe directly below this line that this is caused by the lack of pg_shadow access. I can move the comment up or into the warning statement if that works better.

File: manifests/server/grant.pp (hunk 3 above)
Comment (y-trobinso) [reply manifests/server/grant.pp]: This is a holdover from the original module. I would assume this is to allow the role to be defined independently of the grant (ie, in later reviews), and to specify an ordering when that exists.

File: manifests/server/role.pp (hunk 6 above)
Comment (y-trobinso) [reply manifests/server/role.pp]: I describe directly below this line that this is caused by the lack of pg_shadow access. I can move the comment up or into the warning statement if that works better.

File: manifests/server/grant.pp (hunk 5 above)
Comment (y-trobinso) [reply manifests/server/grant.pp]: I'll reword this. This is being pushed to next review, due to the complexity of scanning all tables while working around https://docs.aws.amazon.com/redshift/latest/dg/c_unsupported-postgresql-functions.html.

**Comment by y-trobinso** [comment]
Updated to reflect the review comments above and several fixes from manual testing in Redshift. This also fixes ALL TABLES IN SCHEMA for groups in its UNLESS clause, by changing the SELECT statement from a singleton to a check for `FALSE = ALL`.

File: manifests/server/grant.pp (hunk 4 above)
Comment (danielpops) [reply manifests/server/grant.pp:381]: I wasn't so worried about the security implications, mostly whether or not it would be considered valid syntax to say `GRANT my_privilege ON my_object_type my_object TO my cool role with spaces`. If your testing shows this is valid, then ship it!

**Review by danielpops — APPROVED** [review]
